File: annotation-processor/src/test/java/me/bristermitten/mittenlib/annotations/integration/extension/CustomDeserializerIntegrationTest.java
package me.bristermitten.mittenlib.annotations.integration.extension;

import static org.assertj.core.api.Assertions.assertThat;

import com.google.inject.*;
import com.google.inject.util.Types;
import me.bristermitten.mittenlib.MittenLibConsumer;
import me.bristermitten.mittenlib.annotations.integration.ConfigLoaderModule;
import me.bristermitten.mittenlib.annotations.integration.extension.fallback.CustomTypeFallback;
import me.bristermitten.mittenlib.annotations.integration.extension.fallback.CustomTypeFallbackConfig;
import me.bristermitten.mittenlib.annotations.integration.extension.fallback.CustomTypeFallbackConfigImpl;
import me.bristermitten.mittenlib.config.ConfigInfrastructureModule;
import me.bristermitten.mittenlib.config.Configuration;
import me.bristermitten.mittenlib.config.DeserializationFunction;
import me.bristermitten.mittenlib.config.SerializationFunction;
import me.bristermitten.mittenlib.config.provider.construct.ConfigProviderFactory;
import me.bristermitten.mittenlib.files.FileTypeModule;
import me.bristermitten.mittenlib.files.yaml.YamlFileType;
import me.bristermitten.mittenlib.watcher.FileWatcherModule;
import org.assertj.core.api.InstanceOfAssertFactories;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class CustomDeserializerIntegrationTest {

    private Injector injector;

    @BeforeEach
    void setup() {
        injector = Guice.createInjector(
                new ConfigInfrastructureModule(),
                new ConfigLoaderModule().asModule(),
                new FileWatcherModule(),
                new FileTypeModule(),
                new AbstractModule() {
                    @Override
                    protected void configure() {
                        bind(MittenLibConsumer.class).toInstance(new MittenLibConsumer("EnumTests"));
                    }
                });
    }

    @Test
    @SuppressWarnings("unchecked")
    void test() {
        var stringReaderProvider = injector.getInstance(ConfigProviderFactory.class)
                .createStringReaderProvider(
                        injector.getInstance(YamlFileType.class),
                        """
                                   customType: 'blahblah'
                                   customTypes: [ 'f' ]
                                """,
                        new Configuration<>(null, CustomTypeConfig.class, CustomTypeConfigImpl.class),
                        (DeserializationFunction<CustomTypeConfig>) injector.getInstance(Key.get(TypeLiteral.get(
                                Types.newParameterizedType(DeserializationFunction.class, CustomTypeConfig.class)))),
                        (SerializationFunction<CustomTypeConfig>) injector.getInstance(Key.get(TypeLiteral.get(
                                Types.newParameterizedType(SerializationFunction.class, CustomTypeConfig.class)))))
                .getOrThrow();

        CustomTypeConfig customTypeConfig = stringReaderProvider.get();
        assertThat(customTypeConfig).extracting(CustomTypeConfig::customType).isEqualTo(new CustomType("hello"));

        assertThat(customTypeConfig)
                .extracting(CustomTypeConfig::customTypes)
                .asInstanceOf(InstanceOfAssertFactories.list(CustomType.class))
                .singleElement()
                .isEqualTo(new CustomType("hello"));
    }

    @Test
    @SuppressWarnings("unchecked")
    void testFallback() {
        var stringReaderProvider = injector.getInstance(ConfigProviderFactory.class)
                .createStringReaderProvider(
                        injector.getInstance(YamlFileType.class),
                        "customType: { test: blahblah }",
                        new Configuration<>(null, CustomTypeFallbackConfig.class, CustomTypeFallbackConfigImpl.class),
                        (DeserializationFunction<CustomTypeFallbackConfig>)
                                injector.getInstance(Key.get(TypeLiteral.get(Types.newParameterizedType(
                                        DeserializationFunction.class, CustomTypeFallbackConfig.class)))),
                        (SerializationFunction<CustomTypeFallbackConfig>)
                                injector.getInstance(Key.get(TypeLiteral.get(Types.newParameterizedType(
                                        SerializationFunction.class, CustomTypeFallbackConfig.class)))))
                .getOrThrow();

        var customTypeConfig = stringReaderProvider.get();
        assertThat(customTypeConfig)
                .extracting(CustomTypeFallbackConfig::customType)
                .extracting(CustomTypeFallback::test)
                .isEqualTo("blahblah");
    }
}

File: annotation-processor/src/test/scala/me/bristermitten/mittenlib/annotations/integration/extension/CustomDeserializerIntegrationTest.scala
package me.bristermitten.mittenlib.annotations.integration.extension

import com.google.inject.{AbstractModule, Guice, Injector, Key, TypeLiteral}
import com.google.inject.util.Types
import me.bristermitten.mittenlib.MittenLibConsumer
import me.bristermitten.mittenlib.annotations.integration.ConfigLoaderModule
import me.bristermitten.mittenlib.annotations.integration.extension.fallback.{
  CustomTypeFallback,
  CustomTypeFallbackConfig,
  CustomTypeFallbackConfigImpl
}
import me.bristermitten.mittenlib.config.{
  ConfigInfrastructureModule,
  Configuration,
  DeserializationFunction,
  SerializationFunction
}
import me.bristermitten.mittenlib.config.provider.construct.ConfigProviderFactory
import me.bristermitten.mittenlib.files.FileTypeModule
import me.bristermitten.mittenlib.files.yaml.YamlFileType
import me.bristermitten.mittenlib.watcher.FileWatcherModule
import org.assertj.core.api.Assertions.assertThat
import org.scalatest.BeforeAndAfterEach
import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.matchers.should.Matchers

class CustomDeserializerIntegrationTest
    extends AnyFunSuite
    with Matchers
    with BeforeAndAfterEach {

  private var injector: Injector = _

  override def beforeEach(): Unit = {
    injector = Guice.createInjector(
      new ConfigInfrastructureModule(),
      new ConfigLoaderModule().asModule(),
      new FileWatcherModule(),
      new FileTypeModule(),
      new AbstractModule() {
        override def configure(): Unit = {
          bind(classOf[MittenLibConsumer]).toInstance(
            new MittenLibConsumer("EnumTests")
          )
        }
      }
    )
  }

  test("test") {
    val deserializerKey = Key
      .get(
        TypeLiteral.get(
          Types.newParameterizedType(
            classOf[DeserializationFunction[_]],
            classOf[CustomTypeConfig]
          )
        )
      )
      .asInstanceOf[Key[DeserializationFunction[CustomTypeConfig]]]

    val serializerKey = Key
      .get(
        TypeLiteral.get(
          Types.newParameterizedType(
            classOf[SerializationFunction[_]],
            classOf[CustomTypeConfig]
          )
        )
      )
      .asInstanceOf[Key[SerializationFunction[CustomTypeConfig]]]

    val stringReaderProvider = injector
      .getInstance(classOf[ConfigProviderFactory])
      .createStringReaderProvider(
        injector.getInstance(classOf[YamlFileType]),
        """
          |customType: 'blahblah'
          |customTypes: [ 'f' ]
          |""".stripMargin,
        new Configuration(
          null,
          classOf[CustomTypeConfig],
          classOf[CustomTypeConfigImpl]
        ),
        injector.getInstance(deserializerKey),
        injector.getInstance(serializerKey)
      )
      .getOrThrow()

    val customTypeConfig = stringReaderProvider.get()
    customTypeConfig.customType() shouldBe new CustomType("hello")

    import scala.jdk.CollectionConverters.*
    val list = customTypeConfig.customTypes().asScala.toList
    list should have size 1
    list.head shouldBe new CustomType("hello")
  }

  test("testFallback") {
    val deserializerKey = Key
      .get(
        TypeLiteral.get(
          Types.newParameterizedType(
            classOf[DeserializationFunction[_]],
            classOf[CustomTypeFallbackConfig]
          )
        )
      )
      .asInstanceOf[Key[DeserializationFunction[CustomTypeFallbackConfig]]]

    val serializerKey = Key
      .get(
        TypeLiteral.get(
          Types.newParameterizedType(
            classOf[SerializationFunction[_]],
            classOf[CustomTypeFallbackConfig]
          )
        )
      )
      .asInstanceOf[Key[SerializationFunction[CustomTypeFallbackConfig]]]

    val stringReaderProvider = injector
      .getInstance(classOf[ConfigProviderFactory])
      .createStringReaderProvider(
        injector.getInstance(classOf[YamlFileType]),
        "customType: { test: blahblah }",
        new Configuration(
          null,
          classOf[CustomTypeFallbackConfig],
          classOf[CustomTypeFallbackConfigImpl]
        ),
        injector.getInstance(deserializerKey),
        injector.getInstance(serializerKey)
      )
      .getOrThrow()

    val customTypeConfig = stringReaderProvider.get()
    customTypeConfig.customType().test() shouldBe "blahblah"
  }
}
